README.md
# text_x_arc

The Helper for draws Arc Text on canvas

#

![](x_arc_text.jpg)

Inspired from [flutter_arc_text](https://developers.mews.com/flutter-how-to-draw-text-along-arc/).


If you just need a way to draw text along an arc, just go with [flutter_arc_text](https://pub.dev/packages/flutter_arc_text).


This package was develop for my special use case so maybe it not suits you.

Consider this package if you need:

- draw text along an arc on canvas
- care about start angle and end angle
- remove all letter out of range (from start angle to end angle)
- center align only when the text shorter than the arc
- customize letter spacing
- tracking what happens on the canvas when draw an arc text.
#

## Basic usage
```
import 'package:flutter/material.dart';
import 'package:flutter/widgets.dart';
import 'package:text_x_arc/text_x_arc.dart';

main() {
  runApp(CustomPaint(
    painter: MyPainter(),
  ));
}

class MyPainter extends CustomPainter {
  @override
  void paint(Canvas canvas, Size size) {
    final centerPos = Offset(size.width / 2, size.height / 2);
    TextPainter textPainter = TextPainter(
        textAlign: TextAlign.justify, textDirection: TextDirection.ltr);
    XArcTextDrawer.draw(
        canvas: canvas,
        centerPos: centerPos,
        radius: 200,
        text:
            'Without requirements or design, programming is the art of adding bugs to an empty text file.',
        textStyle: TextStyle(color: Colors.black, fontSize: 20),
        textPainter: textPainter);
  }

  @override
  bool shouldRepaint(covariant CustomPainter oldDelegate) => true;
}

```

Checkout the  [playground](https://github.com/taodo2291/xarctext_flutter/example/lib/main.dart) example for more information.

Feel free to report bugs or suggest any improvement.
#

## Author
Viet Nguyen - taodo2291@gmail.com
#twitter - @taodo2291


CHANGELOG.md
## [0.0.1] - TODO: Add release date.

* First version


example/lib/test.dart
import 'package:flutter/material.dart';
import 'package:flutter/widgets.dart';
import 'package:text_x_arc/text_x_arc.dart';

main() {
  runApp(CustomPaint(
    painter: MyPainter(),
  ));
}

class MyPainter extends CustomPainter {
  @override
  void paint(Canvas canvas, Size size) {
    final centerPos = Offset(size.width / 2, size.height / 2);
    TextPainter textPainter = TextPainter(
        textAlign: TextAlign.justify, textDirection: TextDirection.ltr);
    XArcTextDrawer.draw(
        canvas: canvas,
        centerPos: centerPos,
        radius: 200,
        text:
            'Without requirements or design, programming is the art of adding bugs to an empty text file.',
        textStyle: TextStyle(color: Colors.black, fontSize: 20),
        textPainter: textPainter);
  }

  @override
  bool shouldRepaint(covariant CustomPainter oldDelegate) => true;
}


test/text_x_arc_test.dart
// import 'package:flutter_test/flutter_test.dart';

// import 'package:text_x_arc/text_x_arc.dart';

// void main() {
//   test('adds one to input values', () {
//     final calculator = Calculator();
//     expect(calculator.addOne(2), 3);
//     expect(calculator.addOne(-7), -6);
//     expect(calculator.addOne(0), 1);
//     expect(() => calculator.addOne(null), throwsNoSuchMethodError);
//   });
// }


example/arc_example.dart
import 'package:flutter/material.dart';
import 'package:text_x_arc/text_x_arc.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return CustomPaint(
      painter: MyPainter(),
    );
  }
}

class MyPainter extends CustomPainter {
  @override
  void paint(Canvas canvas, Size size) {
    TextPainter textPainter = TextPainter(
        textAlign: TextAlign.justify, textDirection: TextDirection.ltr);
    XArcTextDrawer.draw(
      canvas: canvas,
      centerPos: Offset(200, 200),
      radius: 100,
      startAngle: 90,
      endAngle: 270,
      textArcDirection: TextArcDirection.AntiClockwise,
      text: 'hôm nay ăn gì thế nhỉ',
      textStyle: TextStyle(color: Colors.white, fontSize: 30),
      baseline: TextArcBaseline.Center,
      watcher: ArcTextWatcher(),
      textPainter: textPainter,
    );
  }

  @override
  bool shouldRepaint(CustomPainter oldDelegate) {
    return true;
  }
}

class ArcTextWatcher extends Watcher {
  @override
  void drawText(InlineSpan text, TextStyle style) {}

  @override
  void rawPath(Path path, int width) {}

  @override
  void restore() {}

  @override
  void rotate(double rotation) {}

  @override
  void save() {}

  @override
  void scale(double scale) {}

  @override
  void translate(double dx, double dy) {}
}


lib/src/x_arc_text_drawer.dart
import 'package:flutter/material.dart';
import 'package:vector_math/vector_math.dart' as v_math;

/// The Text Draw watcher
///
/// Using it for record canvas events happens during rendering arc text
abstract class Watcher {
  /// canvas rotate
  void rotate(double rotation);

  /// canvas scale
  void scale(double scale);

  /// canvas translate
  void translate(double dx, double dy);

  /// canvas restore
  void restore();

  ///canvas save
  void save();

  ///canvas draw text
  void drawText(String text, TextStyle style, Offset position);
}

class _DefaultWatcher extends Watcher {
  @override
  void drawText(String text, TextStyle style, Offset position) {}

  @override
  void restore() {}

  @override
  void rotate(double rotation) {}

  @override
  void save() {}

  @override
  void scale(double scale) {}

  @override
  void translate(double dx, double dy) {}
}

class XArcTextDrawer {
  /// Draw arc text
  ///
  ///  Parameters:
  /// * canvas [Canvas] -  the canvas
  /// * centerPos [Offset] -  the center position of the arc
  /// * radius [double] - the radius of the arc
  /// * startAngle [double] - the start text angle in degree format (360)
  /// * endAngle [double] - the end text angle in degree format (360)
  /// * textArcDirection[ArcTextDirection] - the direction of the text clockwise or anti-clockwise
  /// * text [String]  - the display text
  /// * textStyle [TextStyle]
  /// * baseline [ArcTextBaseline] - [TextArcBaseline.Inner, TextArcBaseline.Outer, TextArcBaseline.Center] - how text display relative to the arc
  /// * textPainter [TextPainter] - the text painter
  /// * watcher [Watcher] - the canvas watcher
  /// * letterSpacing - the spacing between letters
  static void draw(
      {required Canvas canvas,
      required Offset centerPos,
      required double radius,
      required String text,
      required TextStyle textStyle,
      required TextPainter textPainter,
      double startAngle = 90,
      double endAngle = 90,
      ArcTextDirection textArcDirection = ArcTextDirection.Clockwise,
      ArcTextBaseline baseline = ArcTextBaseline.Outer,
      Watcher? watcher,
      int letterSpacing = 0}) {
    watcher ??= _DefaultWatcher();

    canvas.save();
    watcher.save();

    canvas.translate(centerPos.dx, centerPos.dy);
    watcher.translate(centerPos.dx, centerPos.dy);

    int multiplyDirection =
        textArcDirection == ArcTextDirection.Clockwise ? 1 : -1;

    double deltaAngle = _calcDeltaAngle(startAngle, endAngle, textArcDirection);
    double halfAngle = deltaAngle / 2;

    textPainter.text = TextSpan(text: text, style: textStyle);
    textPainter.layout();
    double textWidth = textPainter.size.width + (letterSpacing * text.length);
    double textHeight = textPainter.size.height;

    double arcAngleByText = v_math.degrees(textWidth / radius);
    double actualStartAngle;
    double maxAngle = deltaAngle;
    if (deltaAngle > arcAngleByText) {
      actualStartAngle = startAngle +
          multiplyDirection * halfAngle -
          multiplyDirection * arcAngleByText / 2;
    } else {
      actualStartAngle = startAngle;
    }

    if (baseline == ArcTextBaseline.Outer) {
      radius += textHeight / 2;
    } else if (baseline == ArcTextBaseline.Inner) {
      radius -= textHeight / 2;
    }

    Offset characterOffset =
        Offset(0, -radius * multiplyDirection - textHeight / 2);
    final rotationInRadian =
        v_math.radians(actualStartAngle + multiplyDirection * 90);
    canvas.rotate(rotationInRadian);
    watcher.rotate(rotationInRadian);

    double angle = 0;
    double prevIncreaseAngle = 0;

    for (String character in text.characters) {
      textPainter.text = TextSpan(text: character, style: textStyle);
      textPainter.layout();
      final characterWidth = textPainter.size.width + letterSpacing;
      final characterAngle = characterWidth / radius;

      angle += v_math.degrees(characterAngle);
      if (angle > maxAngle) {
        break;
      }
      canvas.rotate(prevIncreaseAngle);
      watcher.rotate(prevIncreaseAngle);

      prevIncreaseAngle = multiplyDirection * characterAngle;
      textPainter.paint(canvas, characterOffset);

      watcher.drawText(character, textStyle, characterOffset);
    }
    canvas.restore();
    watcher.restore();
  }

  static double _calcDeltaAngle(
      double startAngle, double endAngle, ArcTextDirection textArcDirection) {
    if (textArcDirection == ArcTextDirection.Clockwise) {
      if (endAngle <= startAngle) {
        return 360 - (startAngle - endAngle);
      }
      return endAngle - startAngle;
    } else {
      if (startAngle > endAngle) {
        return startAngle - endAngle;
      }
      return 360 - (endAngle - startAngle);
    }
  }
}

/// Text Arc Base Line
/// The base line of the text with the circle
enum ArcTextBaseline { Outer, Inner, Center }

/// The Arc text direction
enum ArcTextDirection { Clockwise, AntiClockwise }


example/lib/main.dart
import 'package:dashbook/dashbook.dart';
import 'package:flutter/material.dart';

import 'package:text_x_arc/text_x_arc.dart';

void main() {
  final dashbook = Dashbook(
    theme: ThemeData(),
  );
  dashbook.storiesOf('XArtTextDrawer').decorator(CenterDecorator()).add(
        'Arc Text',
        (context) => Container(
          width: 300,
          child: CustomPaint(
            painter: ArcTextPainter(
              ArcTextExampleModel(
                startAngle: context.numberProperty('start Angle(degree)', 90),
                endAngle: context.numberProperty('end Angle(degree)', 90),
                baseline: context.listProperty('text baseline',
                    ArcTextBaseline.Outer, ArcTextBaseline.values),
                letterSpacing:
                    context.numberProperty('letter spacing', 0.0).toInt(),
                radius: context.numberProperty('radius', 200),
                text: context.textProperty('text',
                    'Without requirements or design, programming is the art of adding bugs to an empty text file.'),
                textDirection: context.listProperty('text direction',
                    ArcTextDirection.Clockwise, ArcTextDirection.values),
                drawCircle: context.boolProperty('draw circle', true),
                style: TextStyle(
                    backgroundColor:
                        context.colorProperty('background', Colors.white),
                    color: context.colorProperty('text color', Colors.black),
                    fontSize: context.numberProperty('font size', 20),
                    fontStyle: context.listProperty(
                        'font style', FontStyle.normal, FontStyle.values),
                    fontWeight: context.listProperty(
                        'font weight', FontWeight.normal, FontWeight.values)),
              ),
            ),
          ),
        ),
      );
  runApp(dashbook);
}

class ArcTextExampleModel {
  double radius;
  double startAngle;
  double endAngle;
  ArcTextDirection textDirection;
  String text;
  int letterSpacing;
  ArcTextBaseline baseline;
  bool drawCircle;
  TextStyle style;

  ArcTextExampleModel({
    required this.radius,
    required this.startAngle,
    required this.endAngle,
    required this.textDirection,
    required this.text,
    required this.letterSpacing,
    required this.baseline,
    required this.drawCircle,
    required this.style,
  });
}

class ArcTextPainter extends CustomPainter {
  final ArcTextExampleModel model;

  ArcTextPainter(this.model);

  @override
  void paint(Canvas canvas, Size size) {
    final centerPos = Offset(size.width / 2, size.height / 2);
    TextPainter textPainter = TextPainter(
        textAlign: TextAlign.justify, textDirection: TextDirection.ltr);
    XArcTextDrawer.draw(
      canvas: canvas,
      centerPos: Offset(size.width / 2, size.height / 2),
      radius: model.radius,
      startAngle: model.startAngle,
      endAngle: model.endAngle,
      textArcDirection: model.textDirection,
      text: model.text,
      textStyle: model.style,
      baseline: model.baseline,
      textPainter: textPainter,
      letterSpacing: model.letterSpacing,
    );

    if (model.drawCircle) {
      canvas.drawCircle(
          centerPos, model.radius, Paint()..style = PaintingStyle.stroke);
    }
  }

  @override
  bool shouldRepaint(covariant CustomPainter oldDelegate) => true;
}
